### PR TITLE
Implement JIT_NewArr1_R2R wrapper

### PR DESCRIPTION
**Pull request by ruben-ayrapetyan**
Implement JIT_NewArr1_R2R as R2R wrapper for JIT_NewArr1 to support both MethodTable-based and TypeDesc-based helpers.

Related issue: #12463

### CODE REVIEW

**Comment by ruben-ayrapetyan** [comment]
@jkotas, Please, take a look.

**Comment by ruben-ayrapetyan** [comment]
Should a regression test be created for this?

File: src/inc/corinfo.h [hunk 1]
```diff
@@ -400,6 +400,7 @@ enum CorInfoHelpFunc
     CORINFO_HELP_NEW_MDARR,         // multi-dim array helper (with or without lower bounds - dimensions passed in as vararg)
     CORINFO_HELP_NEW_MDARR_NONVARARG,// multi-dim array helper (with or without lower bounds - dimensions passed in as unmanaged array)
     CORINFO_HELP_NEWARR_1_DIRECT,   // helper for any one dimensional array creation
+    CORINFO_HELP_NEWARR_1_R2R_DIRECT, // wrapper for R2R direct call, which extracts method table from ArrayTypeDesc
```
Comment (jkotas) [inline src/inc/corinfo.h]: This needs change of the JIT-EE interface GUID

File: src/jit/importer.cpp [hunk 2]
```diff
@@ -3009,6 +3009,7 @@ GenTreePtr Compiler::impInitializeArrayIntrinsic(CORINFO_SIG_INFO* sig)
         newArrayCall->gtCall.gtCallMethHnd != eeFindHelper(CORINFO_HELP_NEWARR_1_VC) &&
         newArrayCall->gtCall.gtCallMethHnd != eeFindHelper(CORINFO_HELP_NEWARR_1_ALIGN8)
 #ifdef FEATURE_READYTORUN_COMPILER
+        && newArrayCall->gtCall.gtCallMethHnd != eeFindHelper(CORINFO_HELP_NEWARR_1_R2R_DIRECT)
```
Comment (jkotas) [inline src/jit/importer.cpp]: Similar change should be made also in utils.cpp and valuenum.cpp

File: src/zap/zapinfo.cpp [hunk 3]
```diff
@@ -3378,7 +3378,7 @@ CorInfoHelpFunc ZapInfo::getCastingHelper(CORINFO_RESOLVED_TOKEN * pResolvedToke
 CorInfoHelpFunc ZapInfo::getNewArrHelper(CORINFO_CLASS_HANDLE arrayCls)
 {
 	if (IsReadyToRunCompilation())
-		return CORINFO_HELP_NEWARR_1_DIRECT;
+		return CORINFO_HELP_NEWARR_1_R2R_DIRECT;
```
Comment (jkotas) [inline src/zap/zapinfo.cpp]: Appending of `ELEMENT_TYPE_NATIVE_ARRAY_TEMPLATE_ZAPSIG` in ComputeRuntimeLookupForSharedGenericToken should be skipped for `IsReadyToRunCompilation()` now

**Comment by jkotas** [comment]
```
05:18:14 There were errors in formatting. Please run jit-format locally with: 
05:18:14 
05:18:14 	jit-format -a x64 -b Checked -o Windows_NT -c <absolute-path-to-coreclr> --verbose --fix --projects dll
05:18:14 
05:18:14 
05:18:14 Or download and apply generated patch:
05:18:14 wget https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_windows_nt_formatting_prtest/8062/artifact/format.patch
05:18:14 git apply format.patch
```

**Comment by jkotas** [comment]
> Should a regression test be created for this?

We do not have the right infrastructure for it, but it would be definitely nice to have better coverage for the R2R native image compatibility with different runtimes. It would need an infrastructure that creates or downloads R2R native images tests compiled against the old runtime, and then runs them against the current runtime. Worth filing an issue at least.

**Comment by jkotas** [comment]
cc @briansull

**Comment by ruben-ayrapetyan** [comment]
> We do not have the right infrastructure for it, but it would be definitely nice to have better coverage for the R2R native image compatibility with different runtimes.

The failure occured with same runtime also. Can the `tests/src/readytorun/tests` be right place for the test?

I think of a test based on the following case (it fails on the a4ddcd62923a5b8ea8babf07d0270092414ca8bf):
```
using System;                                                                                                          
using System.Runtime.CompilerServices;                                                                                 
                                                                                                                       
namespace Application                                                                                                  
{                                                                                                                      
    public class Program                                                                                               
    {                                                                                                               
        [MethodImpl(MethodImplOptions.NoInlining)]                                                                     
        static void f<T>()                                                                                           
        {                                                                                                              
            T [] arr = new T[1000];                                                                                 
        }                                                                                                              
                                                                                                                       
        public static void Main(string[] args)                                                                         
        {                                                                                                              
            f<object>();                                                                                             
        }                                                                                                              
    }                                                                                                                  
}
```

File: src/inc/corinfo.h (hunk 1 above)
Comment (ruben-ayrapetyan) [inline src/inc/corinfo.h]: Fixed.

File: src/zap/zapinfo.cpp (hunk 3 above)
Comment (ruben-ayrapetyan) [inline src/zap/zapinfo.cpp]: Fixed.

File: src/jit/importer.cpp (hunk 2 above)
Comment (ruben-ayrapetyan) [inline src/jit/importer.cpp]: Fixed.

**Comment by jkotas** [comment]
> The failure occured with same runtime also. 

Ah ok .... I thought it is cross-runtime only.

> Can the tests/src/readytorun/tests be right place for the test?

Yes.

**Comment by ruben-ayrapetyan** [comment]
The "Windows_NT x86 Checked Build and Test" should be fixed by #12481.

**Comment by ruben-ayrapetyan** [comment]
>> Can the tests/src/readytorun/tests be right place for the test?

>Yes.

Added `newarray.cs`

**Comment by BruceForstall** [comment]
cc @dotnet/jit-contrib : this PR will change the JIT-EE interface version

**Pull request merged**